src/torchcodec/samplers/_implem.py

```diff
@@ -0,0 +1,62 @@
+from typing import List
+
+import torch
+
+from torchcodec.decoders import (  # TODO: move FrameBatch to torchcodec.FrameBatch?
+    FrameBatch,
+    SimpleVideoDecoder,
+)
+
+
+def clips_at_random_indices(
+    decoder: SimpleVideoDecoder,
+    *,
+    num_clips: int = 1,
+    num_frames_per_clip: int = 1,
+    num_indices_between_frames: int = 1,
+) -> List[FrameBatch]:
+    if num_clips <= 0:
+        raise ValueError(f"num_clips ({num_clips}) must be strictly positive")
+    if num_frames_per_clip <= 0:
+        raise ValueError(
+            f"num_frames_per_clip ({num_frames_per_clip}) must be strictly positive"
+        )
+    if num_indices_between_frames <= 0:
+        raise ValueError(
+            f"num_indices_between_frames ({num_indices_between_frames}) must be strictly positive"
+        )
+
+    # Determine the span of a clip, i.e. the number of frames (or indices)
+    # between the first and last frame in the clip, both included. This isn't
+    # the same as the number of frames in a clip!
+    # Example: f means a frame in the clip, x means a frame excluded from the clip
+    # num_frames_per_clip = 4
+    # num_indices_between_frames = 1, clip = ffff      , span = 4
+    # num_indices_between_frames = 2, clip = fxfxfxf   , span = 7
+    # num_indices_between_frames = 3, clip = fxxfxxfxxf, span = 10
+    clip_span = num_indices_between_frames * (num_frames_per_clip - 1) + 1
+
+    # TODO: We should probably not error.
+    if clip_span > len(decoder):
+        raise ValueError(
+            f"Clip span ({clip_span}) is larger than the number of frames ({len(decoder)})"
+        )
+
+    last_clip_start_index = len(decoder) - clip_span
+    clip_start_indices = torch.randint(
+        low=0, high=last_clip_start_index + 1, size=(num_clips,)
+    )
+
```

Comment:
(I'm putting the comment here so that we can still read the logic without a giant textbox in the middle of it.)

I'm not sure if we want to compute and then use `last_clip_start_index`. From a probability perspective, I think we're actually making it much less likely that the frames in the range `[last_clip_start_index, len(decoder))` are ever returned to the user. For all other frames, consider that they could be the start of a clip, or included in a clip. For last set of frames, they're only included if we happen to select `last_clip_start_index` as a clip start.

An alternative approach is that we don't do this, and just set `low=0, high=len(decoder)`. Then we let our clip-too-small policy from above handle the case when the clip start is is in the range `len(decoder) - clip_span`.

Both options result in some quirky behavior. What is currently implemented makes it less likely (I think) that we'll ever return frames in the range `[last_clip_start_index, len(decoder))`. My suggestion above means that selecting those frames is just as likely as the frames in the rest of the video, but we'll always have some duplicates - and maybe biasing the earlier part of the video is better than having duplicates. I'm not sure, sampling is actually a hard problem. `¯\_(ツ)_/¯`

Reply:
I think uniformly sampling the clip_starts in the range where they are valid matches my own intuition.

Duplicate frames may not be good for training.

If @scotts has a strong opinion, maybe we can expose an option but the current behavior seems good to me as the default.

Reply:
After chatting offline we concluded the best way to handle this was to let the user choose the sampling range, similarly to the "equaly_spaced_in_time" sampler.